### PR TITLE
iblock read-only device changes

### PR DESCRIPTION
**Pull request by ddiss**

### CODE REVIEW

**Comment by agrover** [comment]
looks good to me. @mikechristie @gonzoleeman @maurizio-lombardi any objections?

File: targetcli/ui_backstore.py [hunk 1]
```diff
@@ -526,17 +529,45 @@ def __init__(self, parent):
...
+            readonly = self._ui_block_ro_check(dev)
+        else:
+            readonly = self.ui_eval_param(readonly, 'bool', False)
+
         wwn = self.ui_eval_param(wwn, 'string', None)
 
-        so = BlockStorageObject(name, dev, readonly=readonly, wwn=wwn)
+        try:
+            so = BlockStorageObject(name, dev, readonly=readonly, wwn=wwn)
+        except IOError:
```
Comment (mikechristie) [inline targetcli/ui_backstore.py]: It was not clear why this change is needed. Which call in the BlockStorageObject constructor raises it? Is it set_udev_path?

Why also only catch this one here and not also catch the possible RTSLibError here too or have the higher level catch it with all exceptions?

**Comment by gonzoleeman** [comment]
I'm interested in the answer to Mike's questions, but otherwise have no objection.

File: targetcli/ui_backstore.py (hunk 1 above)
Comment (ddiss) [reply targetcli/ui_backstore.py]: It's the rtslib backstore->enable I/O that triggers IOError: [Errno 13]
```
  File "/usr/lib/python2.7/site-packages/targetcli/ui_backstore.py", line 557, in ui_command_create
    so = BlockStorageObject(name, dev, readonly=readonly, wwn=wwn)
  File "/usr/lib/python2.7/site-packages/rtslib_fb/tcm.py", line 703, in __init__
    self._configure(dev, wwn, readonly)
  File "/usr/lib/python2.7/site-packages/rtslib_fb/tcm.py", line 720, in _configure
    self._enable()
  File "/usr/lib/python2.7/site-packages/rtslib_fb/tcm.py", line 158, in _enable
    fwrite(path, "1\n")
  File "/usr/lib/python2.7/site-packages/rtslib_fb/utils.py", line 79, in fwrite
    file_fd.write(str(string))
IOError: [Errno 13] Permission denied
```

Even with the patch to auto-detect block device read-only state, this can still be tripped if a read-only block device is provided alongside an explicit readonly=false, hence the need for this exception handler.

I'm not aware of other exceptions that can occur in this code path, so chose to only cater to the above IOError. I could change this to something more generic if you'd prefer.

Thanks everyone for the feedback so far.

Comment (mikechristie) [reply targetcli/ui_backstore.py]: > I'm not aware of other exceptions that can occur in this code path, so chose to only cater to the above IOError. I could change this to something more generic if you'd prefer.

Every kernel backend module can return failure from their device configure callback so a write to the enable file can return failure from any of them, so I think we want to handle this in one place and not just fix one backend.

For example maybe the StorageObject class's _enable function can catch these errors and translate them to RTSLibErrors which are being caught already.
 
I think having one or two backend create functions catching it and the rest crashing is the worst option :) My patch
6b14c2af6790ad8073fcd4982363197f5c030de2
was the wrong way to go :) It should be reverted and use the generic code needed in this thread. I goofed up and did not know the other constrictors did not catch these errors :)

Comment (ddiss) [reply targetcli/ui_backstore.py]: Okay, I'll look at adding generic RTSLibErrors functionality. That said, I think the first commit could go in without the follow-up error handling change.

Comment (mikechristie) [reply targetcli/ui_backstore.py]: Agree.

**Comment by ddiss** [comment]
This new version drops the error handling change (to be reworked & submitted separately), and adds only the read-only autodetect logic.

**Comment by mikechristie** [comment]
Looks good to me.

**Comment by agrover** [comment]
thanks!!
